### DISCUSSION TITLE
To improve GlycanTreeModeler computing efficiency

### DISCUSSION THREAD

**Discussion by Venthaul** [RosettaCommons/rosetta#274] [RosettaScripts]
I am constructing a protein system with multiple glycosylation sites using GlycanTreeModeler. Use the following code:
```
    <MOVERS>
        <GlycanTreeModeler name="model" quench_mode="false" rounds="1" layer_size="2" window_size="1" />
    </MOVERS>
    <PROTOCOLS>
        <Add mover="model"/>
    </PROTOCOLS>
```
In my subsequent plans, I will use molecular dynamics simulations, so I can tolerate relatively low computational accuracy. However, such code still took about a week to complete the glycan tree optimization. 

Which parameter can improve the speed of GlycanTreeModeler? Thank you!

**Comment by roccomoretti** [comment]
I'm not all that familiar with the GlycanTreeModeler, but from a glance at the code and the [documentation](https://docs.rosettacommons.org/docs/latest/scripting_documentation/RosettaScripts/Movers/movers_pages/carbohydrates/GlycanTreeModeler), here's some parameters which might affect runtime:

* quench_mode (default false) -- instead modeling all glycans simultaneously, optimize them one by one in a random order
* rounds (default 1) -- if set to more than 1, this may increase the time needed (doesn't look like an issue in your case)
* hybrid_protocol (default true) -- if true, do a more extensive optimization each time you add a glycan to the growing chain.
* final_min_pack_min (default true) -- if true, do a final optimization of the structure
* glycan_sampler_rounds (default 100) -- the number of rounds in the internal glycan modeling loop

I'd probably recommend trying `quench_mode` first, and seeing how that affects the output. Next I'd probably look into turning off hybrid_protocol. Turning off final_min_pack_min might be worth it in your case (as you're going on to additional optimization anyway), though it might not actually contribute much to the runtime. Turning down the glycan_sampler_rounds might also give you a speed boost, but the effect will likely be marginal unless you greatly reduce it, at which point you're likely going to affect the quality of the output more.

**Reply by Venthaul** [reply]
Thanks for your suggestions! I will try these.

**Reply by Venthaul** [reply]
Hello, quench_mode=true may lead to segmentation fault. 

```
[END_BACKTRACE]

[FILE]: SIGSEGV
[LINE]: 11
[START_MESSAGE]
Segmentation Fault

[END_MESSAGE]
[END_CRASH_REPORT]



core.pack.pack_rotamers: (0) built 0 rotamers at 0 positions.
core.pack.interaction_graph.interaction_graph_factory: (0) Instantiating DensePDInteractionGraph
basic.thread_manager.RosettaThreadManager: (0) [ WARNING ] A work vector of size zero was passed to the RosettaThreadManager!  Duly returning without doing anything.
protocols.carbohydrates.util: (0) pack accepted: 1
protocols.carbohydrates.util: (0) Shear accepts: 40/40
protocols.carbohydrates.util: (0) min accept: 1
core.pack.task: (0) Packer task: initialize from command line() 
core.pack.pack_rotamers: (0) built 0 rotamers at 0 positions.
core.pack.interaction_graph.interaction_graph_factory: (0) Instantiating DensePDInteractionGraph
basic.thread_manager.RosettaThreadManager: (0) [ WARNING ] A work vector of size zero was passed to the RosettaThreadManager!  Duly returning without doing anything.
protocols.carbohydrates.util: (0) pack accepted: 1
protocols.carbohydrates.GlycanSampler: (0) energy final post min: 2463.64
protocols.carbohydrates.GlycanSampler: (0) energy recov post min: 2463.64
protocols.moves.TrialCounter: (0)              unk trials=    254;  accepts= 0.9134;  energy_drop/trial=  -0.31165
protocols.carbohydrates.GlycanTreeModeler: (0) De-Virtualizing current foliage layer
protocols.carbohydrates.GlycanTreeModeler: (0) Running the GlycanSampler on layer [ start -> end (including) ]: 1 1
protocols.carbohydrates.GlycanTreeModeler: (0) Applying normal protocol
protocols.carbohydrates.GlycanSampler: (0) initializing objects 
protocols.carbohydrates.GlycanSampler: (0) Modeling 2 glycan residues
protocols.carbohydrates.GlycanSampler: (0) Initialized
protocols.carbohydrates.GlycanSampler: (0) starting energy: 5928.65
protocols.carbohydrates.GlycanSampler: (0) Total Rounds = 372 ( 2 glycans * 186 )
protocols.carbohydrates.LinkageConformerMover: (0) Sampling alpha-L-Fucp-(?->6)-beta-D-GlcpNAc- linkage 
protocols.carbohydrates.LinkageConformerMover: (0) Sampling conformer 1 which has a population of 0.8187


AN INTERNAL ERROR HAS OCCURED. PLEASE SEE THE CONTENTS OF ROSETTA_CRASH.log FOR DETAILS.


Segmentation fault (core dumped)
```
